src/Microsoft.Diagnostics.Runtime/src/ImmutableArrayExtension.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+
+namespace Microsoft.Diagnostics.Runtime
+{
+    internal static class ImmutableArrayExtension
+    {
+        /// <summary>
+        /// <b>WARNING:</b> Typing this method may cause Visual Studio IntelliCode 2.2 to crash.
+        /// </summary>
+        internal static unsafe Span<T> DangerousGetSpan<T>(this ImmutableArray<T>.Builder builder) where T : unmanaged
+        {
+            // MemoryMarshal.CreateSpan isn't available on .NET Standard 2.0
+            fixed (T* pointer = &builder.ItemRef(0))
```

Comment:
I don't think this is correct.  We are fixing a pointer to an immutable array then returning it as a Span.  If the GC decides to relocate this poitner   The memory backed by ImmutableArray<T>.Builder.ItemRef(0) could be relocated by the GC at any time after returning, leading to heap corruption if we try to write to it after that.  (Even if the runtime _can_ track this pointer, I still think I'd like to avoid this pattern.  It looks wrong from code review, and I'd be hard pressed to prove its correctness.)

Since we only use this method in one place around creating the buildid, I'd rather just be inefficient at creating the buildid...  Stackalloc the byte[] to get the buildid, then copy it over to an ImmutableArray, it's ok if we have to be a little inefficient when it makes sense to.

Reply:
Well, that unmanaged pointer passed to the span will be stored as a managed pointer, and since any managed pointer is tracked by GC it can by updated whenever the array is relocated. And the span will always be valid because the array doesn't come from array pool and therefore won't be modified for reuse. So this method is quite safe...

If you don't like this idea, what about a hack that directly converts an array into an immutable array?

If you don't like both... I'll take copying instead.

Reply:
A coworker pointed out that since I ship two copies of the library anyway, we might as well change the netstandard 2.0 one over to .Net Core 2.1.  Then we could just use MemoryMarshal.CreateSpan and avoid this problem altogether.  Let me make a PR for that real quick and then you can just use CreateSpan.

(Also I'm fine with hacks in general if they get around stuff like this, so I'd be fine with one that converts an array to an immutable one, but I think we can avoid both by updating the library to target .Net Core 2.1 instead of netstandard.)

Reply:
https://github.com/microsoft/clrmd/pull/426.

I'm going to let the CI builds finish to make sure this doesn't break anything unexpected in the weird CI build environment.

Reply:
Sigh.  After all of that I forgot MemoryMarshal.CreateSpan isn't in 4.7.1, so it's a moot point.

Go for the hack that converts an array into an immutable array.

Reply:
It's odd that System.Memory.dll doesn't have MemoryMarshal.CreateSpan.  I'm not sure why that's the case, but I doubt it's an oversight.  It might just be that it can't be supported on desktop CLR due to the implementation.

Reply:
It it in .NET Standard 2.1 that .NET Framework won't support. Nothing to do with the implementation...

Reply:
66a93ff seems fairly reasonable to me.  Let me know when this is ready for review/merge.

Reply:
This is ready for merge. Further updates (if any) can be made in separate PRs.

Reply:
Done.

Reply:
FYI `MemoryMarshal.CreateSpan` needs runtime support, but the `System.Memory` package targets .NET Standard and is not exclusive for CLRs.

From dotnet/corefx# 32943:

> MemoryMarshal.CreateSpan is the new method name, and is available in .NET Core 2.1. It's not available in the portable implementation because it's not implementable safely without runtime support (it would have similar problems to what you cite regarding passing around pointers).